src/work/BasicWork.cpp

```diff
@@ -0,0 +1,263 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "work/BasicWork.h"
+#include "lib/util/format.h"
+#include "util/Logging.h"
+#include "util/Math.h"
+
+namespace stellar
+{
+
+size_t const BasicWork::RETRY_NEVER = 0;
+size_t const BasicWork::RETRY_ONCE = 1;
+size_t const BasicWork::RETRY_A_FEW = 5;
+size_t const BasicWork::RETRY_A_LOT = 32;
+size_t const BasicWork::RETRY_FOREVER = 0xffffffff;
+
+BasicWork::BasicWork(Application& app, std::function<void()> callback,
+                     std::string name, size_t maxRetries)
+    : mApp(app)
+    , mNotifyCallback(std::move(callback))
+    , mName(std::move(name))
+    , mMaxRetries(maxRetries)
+{
+    reset();
+}
+
+BasicWork::~BasicWork()
+{
+    mState = WORK_DESTRUCTING;
```

Comment:
At this point, lifetime of BasicWork already ended, so no external entity should be able to access mState.
I see it is only used in wakeUp which is called mainly from exit.async_wait() calls.

I think we only execute asio jobs in main thread, so it looks like it may not be possible to call wakeUp after destructor is called (as we always use weak_ptrs to Work items in such cases).

Reply:
@vogel is right: setting to `WORK_DESTRUCTING` only works when doing it from a derived class, so you probably need a special method to do just that (typically people call it `shutdown`)